<a href="https://colab.research.google.com/github/phurisk/BSC_DPDM23/blob/main/Frequent_Patterns_(Association_Rules).ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [1]:
import pandas as pd
import os

In [2]:
from google.colab import drive
drive.mount('/content/drive')
path = '/content/drive/MyDrive/BSC_DPDM23/File'

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [97]:
data = pd.read_csv(os.path.join(path,'reduced_marketbasket.csv'),encoding='latin-1')

In [4]:
data

InvoiceNo StockCode                          Description  Quantity  \
0         536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1         536365     71053                  WHITE METAL LANTERN         6   
2         536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3         536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4         536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   
...          ...       ...                                  ...       ...   
541904    581587     22613          PACK OF 20 SPACEBOY NAPKINS        12   
541905    581587     22899         CHILDREN'S APRON DOLLY GIRL          6   
541906    581587     23254        CHILDRENS CUTLERY DOLLY GIRL          4   
541907    581587     23255      CHILDRENS CUTLERY CIRCUS PARADE         4   
541908    581587     22138        BAKING SET 9 PIECE RETROSPOT          3   

             InvoiceDate  UnitPrice  CustomerID         Country  
0       12/01/2010 08:26       2.55     17850.0  United Kingdom  
1       12/01/2010 08:26       3.39     17850.0  United Kingdom  
2       12/01/2010 08:26       2.75     17850.0  United Kingdom  
3       12/01/2010 08:26       3.39     17850.0  United Kingdom  
4       12/01/2010 08:26       3.39     17850.0  United Kingdom  
...                  ...        ...         ...             ...  
541904  12/09/2011 12:50       0.85     12680.0          France  
541905  12/09/2011 12:50       2.10     12680.0          France  
541906  12/09/2011 12:50       4.15     12680.0          France  
541907  12/09/2011 12:50       4.15     12680.0          France  
541908  12/09/2011 12:50       4.95     12680.0          France  

[541909 rows x 8 columns]

## แปลง data ให้อยู่ในรูปแบบ transaction data ด้วย group by .appy(list)

In [5]:
data.groupby('InvoiceNo')['StockCode'].apply(list)

InvoiceNo
536365     [85123A, 71053, 84406B, 84029G, 84029E, 22752,...
536366                                        [22633, 22632]
536367     [84879, 22745, 22748, 22749, 22310, 84969, 226...
536368                          [22960, 22913, 22912, 22914]
536369                                               [21756]
                                 ...                        
C581484                                              [23843]
C581490                                       [22178, 23144]
C581499                                                  [M]
C581568                                              [21258]
C581569                                       [84978, 20979]
Name: StockCode, Length: 25900, dtype: object

In [6]:
transactions = data.groupby('InvoiceNo')['StockCode'].apply(list)

- ดูประเทศ

In [ ]:
set(data["Country"])

## Association Rules

In [8]:
!pip install apyori

In [9]:
from apyori import apriori

In [134]:
frequentItemsets = list(apriori(transactions,min_support = 0.065))
frequentItemsets

[RelationRecord(items=frozenset({'CHOCOLATE HOT WATER BOTTLE'}), support=0.07308641975308643, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'CHOCOLATE HOT WATER BOTTLE'}), confidence=0.07308641975308643, lift=1.0)]),
 RelationRecord(items=frozenset({'HAND WARMER BABUSHKA DESIGN'}), support=0.08148148148148149, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'HAND WARMER BABUSHKA DESIGN'}), confidence=0.08148148148148149, lift=1.0)]),
 RelationRecord(items=frozenset({'HEART OF WICKER SMALL'}), support=0.07012345679012345, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'HEART OF WICKER SMALL'}), confidence=0.07012345679012345, lift=1.0)]),
 RelationRecord(items=frozenset({'HOT WATER BOTTLE BABUSHKA '}), support=0.06567901234567901, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'HOT WATER BOTTLE BABUSHKA '}), confidence=0.06567901234567901, lift=1.0)

In [11]:
len(frequentItemsets)

4

In [12]:
frequentItemsets[0][0]

frozenset({'22423'})

In [13]:
dir(frequentItemsets[0][0])

['__and__',
 '__class__',
 '__class_getitem__',
 '__contains__',
 '__delattr__',
 '__dir__',
 '__doc__',
 '__eq__',
 '__format__',
 '__ge__',
 '__getattribute__',
 '__gt__',
 '__hash__',
 '__init__',
 '__init_subclass__',
 '__iter__',
 '__le__',
 '__len__',
 '__lt__',
 '__ne__',
 '__new__',
 '__or__',
 '__rand__',
 '__reduce__',
 '__reduce_ex__',
 '__repr__',
 '__ror__',
 '__rsub__',
 '__rxor__',
 '__setattr__',
 '__sizeof__',
 '__str__',
 '__sub__',
 '__subclasshook__',
 '__xor__',
 'copy',
 'difference',
 'intersection',
 'isdisjoint',
 'issubset',
 'issuperset',
 'symmetric_difference',
 'union']

In [14]:
frequentItemsets[0][0].__len__()

1

In [15]:
min_sup = 0.03

frequentItemsets = list(apriori(transactions,min_support = min_sup))
for i in frequentItemsets:
    if i[0].__len__()>1:
        print('yes')
        print(i)

yes
RelationRecord(items=frozenset({'22386', '85099B'}), support=0.03216216216216216, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'22386', '85099B'}), confidence=0.03216216216216216, lift=1.0), OrderedStatistic(items_base=frozenset({'22386'}), items_add=frozenset({'85099B'}), confidence=0.6766856214459789, lift=8.20897311262335), OrderedStatistic(items_base=frozenset({'85099B'}), items_add=frozenset({'22386'}), confidence=0.39016393442622943, lift=8.20897311262335)])
yes
RelationRecord(items=frozenset({'22699', '22697'}), support=0.03027027027027027, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'22699', '22697'}), confidence=0.03027027027027027, lift=1.0), OrderedStatistic(items_base=frozenset({'22697'}), items_add=frozenset({'22699'}), confidence=0.7417218543046358, lift=17.1523178807947), OrderedStatistic(items_base=frozenset({'22699'}), items_add=frozenset({'22697'}), confidence=0.7, lift=17.1523178807947)])


RelationRecord(items=frozenset({'22386', '85099B'}),

support=0.03216216216216216,

ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'22386', '85099B'}), confidence=0.03216216216216216, lift=1.0),

OrderedStatistic(items_base=frozenset({'22386'}), items_add=frozenset({'85099B'}), confidence=0.6766856214459789, lift=8.20897311262335),

OrderedStatistic(items_base=frozenset({'85099B'}), items_add=frozenset({'22386'}), confidence=0.39016393442622943, lift=8.20897311262335)])


In [16]:
set(data[(data['StockCode']=='22386') | (data['StockCode']=='85099B') | (data['StockCode']=='22699') | (data['StockCode']=='22697') ]['Description'])

{'GREEN REGENCY TEACUP AND SAUCER',
 'JUMBO BAG PINK POLKADOT',
 'JUMBO BAG RED RETROSPOT',
 'ROSES REGENCY TEACUP AND SAUCER '}

RelationRecord(items=frozenset({'22699', '22697'}), support=0.03027027027027027,

ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'22699', '22697'}), confidence=0.03027027027027027, lift=1.0),

OrderedStatistic(items_base=frozenset({'22697'}), items_add=frozenset({'22699'}), confidence=0.7417218543046358, lift=17.1523178807947),

OrderedStatistic(items_base=frozenset({'22699'}), items_add=frozenset({'22697'}), confidence=0.7, lift=17.1523178807947)])

In [17]:
set(data[(data['StockCode']=='22699') | (data['StockCode']=='22697')]['InvoiceNo'])

{'544574',
 '575846',
 '550989',
 '544657',
 '576329',
 '553387',
 '563505',
 '542102',
 '578067',
 '539609',
 '539492',
 '552694',
 '562534',
 '557431',
 '549730',
 '550251',
 '554124',
 '575324',
 '578994',
 '553870',
 '551850',
 '540981',
 '571216',
 '567663',
 '571546',
 '541280',
 '580501',
 'C545033',
 'C545449',
 '572549',
 '562429',
 '557788',
 '565916',
 'C545525',
 '576618',
 '577298',
 '544928',
 '542221',
 '547788',
 '558066',
 '567336',
 '574334',
 '576202',
 '560590',
 '580983',
 '565155',
 '542590',
 '561125',
 '570701',
 '574731',
 '577512',
 '552270',
 '544778',
 '577258',
 '561329',
 '555499',
 '553546',
 '567639',
 '539442',
 'C562953',
 'C549840',
 '553607',
 '538153',
 '544208',
 '545176',
 '546980',
 '545343',
 '541859',
 '555158',
 '550499',
 '547378',
 '562166',
 '552540',
 '553709',
 '553566',
 '566257',
 '539451',
 '571508',
 '545058',
 '561633',
 '543826',
 '577358',
 '558777',
 '558862',
 '568057',
 '552730',
 '541506',
 '579518',
 '562770',
 '549589',
 '580

In [18]:
data[data['InvoiceNo']=='562950']

InvoiceNo StockCode                         Description  Quantity  \
297488    562950     22752        SET 7 BABUSHKA NESTING BOXES         2   
297489    562950     22755     SMALL PURPLE BABUSHKA NOTEBOOK         12   
297490    562950     22753     SMALL YELLOW BABUSHKA NOTEBOOK         12   
297491    562950     22754        SMALL RED BABUSHKA NOTEBOOK         12   
297492    562950     22757        LARGE RED BABUSHKA NOTEBOOK         12   
297493    562950     22758    LARGE PURPLE BABUSHKA NOTEBOOK          12   
297494    562950     22756     LARGE YELLOW BABUSHKA NOTEBOOK         12   
297495    562950     23310             BUBBLEGUM RING ASSORTED        36   
297496    562950     23280  FOLDING BUTTERFLY MIRROR HOT PINK         12   
297497    562950     22383              LUNCH BAG SUKI DESIGN         10   
297498    562950     22492             MINI PAINT SET VINTAGE         36   
297499    562950     71459      HANGING JAM JAR T-LIGHT HOLDER        12   
297500    562950     84945  MULTI COLOUR SILVER T-LIGHT HOLDER        12   
297501    562950     21175         GIN + TONIC DIET METAL SIGN        12   
297502    562950     23164                  REGENCY CAKE SLICE         4   
297503    562950     23245          SET OF 3 REGENCY CAKE TINS         4   
297504    562950     23171            REGENCY TEA PLATE GREEN         12   
297505    562950     22697     GREEN REGENCY TEACUP AND SAUCER         6   

             InvoiceDate  UnitPrice  CustomerID         Country  
297488  08/11/2011 09:56       8.50     15608.0  United Kingdom  
297489  08/11/2011 09:56       0.85     15608.0  United Kingdom  
297490  08/11/2011 09:56       0.85     15608.0  United Kingdom  
297491  08/11/2011 09:56       0.85     15608.0  United Kingdom  
297492  08/11/2011 09:56       1.25     15608.0  United Kingdom  
297493  08/11/2011 09:56       1.25     15608.0  United Kingdom  
297494  08/11/2011 09:56       1.25     15608.0  United Kingdom  
297495  08/11/2011 09:56       0.42     15608.0  United Kingdom  
297496  08/11/2011 09:56       0.83     15608.0  United Kingdom  
297497  08/11/2011 09:56       1.65     15608.0  United Kingdom  
297498  08/11/2011 09:56       0.65     15608.0  United Kingdom  
297499  08/11/2011 09:56       0.85     15608.0  United Kingdom  
297500  08/11/2011 09:56       0.85     15608.0  United Kingdom  
297501  08/11/2011 09:56       2.55     15608.0  United Kingdom  
297502  08/11/2011 09:56       4.95     15608.0  United Kingdom  
297503  08/11/2011 09:56       4.95     15608.0  United Kingdom  
297504  08/11/2011 09:56       1.65     15608.0  United Kingdom  
297505  08/11/2011 09:56       2.95     15608.0  United Kingdom

## quiz 22386 และ 85099b คือสินค้าอะไร

In [19]:
set(data[(data['StockCode']=='22386')  ]['Description'])

{'JUMBO BAG PINK POLKADOT'}

In [20]:
set(data[(data['StockCode']=='85099B') ]['Description'])

{'JUMBO BAG RED RETROSPOT'}

#Hw หา Frequent Item Set  และ Association Rules
- ของช่วงเวลา Chirtmas & New Year(1 Dec -31 Dec)
- ของประเทศในทวีป Europe
- ของช่วงเวลา ตรุษจีน ในทวีป Asai (15 jan-15 Feb)
- ของช่วงเวลา Chirtmas & New Year ในทวีป Europe (1 Dec -31 Dec)
- ของวันวาเลนไทน์
- โดยจัดกลุ่มตามประเทศ+เดือน

In [295]:
!pip install apyori

In [296]:
from apyori import apriori

In [297]:
df = pd.read_csv(os.path.join(path, 'reduced_marketbasket.csv'), encoding='latin-1')


## ข้อ 1. ช่วงเวลา Chirtmas & New Year(1 Dec -31 Dec)

In [298]:
# Convert the 'InvoiceDate' column to a datetime object
df['InvoiceDate'] = pd.to_datetime(df['InvoiceDate'], format='%m/%d/%Y %H:%M')


In [ ]:
# Filter the data for December 1st to December 31st of all years
filtered_df = df[(df['InvoiceDate'].dt.month == 12) & (df['InvoiceDate'].dt.day >= 1) & (df['InvoiceDate'].dt.day <= 31)]

# Display the filtered data
print(filtered_df)

In [279]:
# transactions= filtered_df.groupby('InvoiceNo')['StockCode'].apply(list)
# transactions  ### ถ้าใช้ transactions แทน transactions_apply จะเป็นการโชว์ StockCode แทนข้อมูลสินค้า ###

In [301]:
transactions = filtered_df.groupby('InvoiceNo')['Description'].apply(list).reset_index() #ทำการ group และทำให้ข้อมูลใน Description เป็นlist
transactions_apply = group['Description'].apply(lambda x: [str(item) for item in x]) #ใช้ฟังก์ชัน lambda เพื่อแปลงสินค้าทุกชิ้นใน list ของ 'transactions' เป็นข้อความ (string)

In [319]:
frequentItemsetsC_and_N = list(apriori(transactions_apply, min_support=0.035))

for record in frequentItemsetsC_and_N:
    if len(record.items) > 1:
        print('yes')
        print(f"Items: {', '.join(record.items)}")
        print(f"Support: {record.support:.4f} ({record.support*100:.2f}%)")

        for i, stat in enumerate(record.ordered_statistics, 1):
            items_base = 'None' if not stat.items_base else ', '.join(stat.items_base)
            items_add = ', '.join(stat.items_add)
            confidence = stat.confidence
            lift = stat.lift

            print(f"Rule {i}: ({items_base} -> {items_add} : Confidence={confidence:.4f}, Lift={lift:.4f})")
        print()


yes
Items: ALARM CLOCK BAKELIKE GREEN, ALARM CLOCK BAKELIKE RED 
Support: 0.0368 (3.68%)
Rule 1: (None -> ALARM CLOCK BAKELIKE GREEN, ALARM CLOCK BAKELIKE RED  : Confidence=0.0368, Lift=1.0000)
Rule 2: (ALARM CLOCK BAKELIKE GREEN -> ALARM CLOCK BAKELIKE RED  : Confidence=0.7089, Lift=14.4627)
Rule 3: (ALARM CLOCK BAKELIKE RED  -> ALARM CLOCK BAKELIKE GREEN : Confidence=0.7517, Lift=14.4627)

yes
Items: HOT WATER BOTTLE I AM SO POORLY, CHOCOLATE HOT WATER BOTTLE
Support: 0.0365 (3.65%)
Rule 1: (None -> HOT WATER BOTTLE I AM SO POORLY, CHOCOLATE HOT WATER BOTTLE : Confidence=0.0365, Lift=1.0000)
Rule 2: (CHOCOLATE HOT WATER BOTTLE -> HOT WATER BOTTLE I AM SO POORLY : Confidence=0.5163, Lift=9.7484)
Rule 3: (HOT WATER BOTTLE I AM SO POORLY -> CHOCOLATE HOT WATER BOTTLE : Confidence=0.6894, Lift=9.7484)

yes
Items: PAPER CHAIN KIT VINTAGE CHRISTMAS, PAPER CHAIN KIT 50'S CHRISTMAS 
Support: 0.0477 (4.77%)
Rule 1: (None -> PAPER CHAIN KIT VINTAGE CHRISTMAS, PAPER CHAIN KIT 50'S CHRISTMAS  : 

สรุป เลือกอันที่มีค่าสนับสนุนมากที่สุด


Items: PAPER CHAIN KIT VINTAGE CHRISTMAS และ PAPER CHAIN KIT 50'S CHRISTMAS ค่าสนับสนุนคือ 0.0477 หรือ 4.77% ของการทำธุรกรรมทั้งหมด

- เมื่อ 'PAPER CHAIN KIT 50'S CHRISTMAS' ถูกซื้อ มีความมั่นใจ 51.60% ว่า 'PAPER CHAIN KIT VINTAGE CHRISTMAS' จะถูกซื้อด้วย

- เมื่อ 'PAPER CHAIN KIT VINTAGE CHRISTMAS' ถูกซื้อ มีโอกาส 66.51% ว่า 'PAPER CHAIN KIT 50'S CHRISTMAS' จะถูกซื้อด้วย